Add support for multiple subscriptions

ryanr14 opened the pull request:
### All Submissions:


* [x] Does your code follow the <a href="https://github.com/woocommerce/woocommerce-sniffs/">WooCommerce Sniffs</a> variant of WordPress coding standards?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?
* [x] Will this change require new documentation or changes to existing documentation?

---

### Changes proposed in this Pull Request:
This would add support for multiple subscriptions in one purchase via Payfast.
Closes <a href="https://linear.app/a8c/issue/PAYFAST-8/can-payfast-support-multiple-subscriptions">PAYFAST-8: Can Payfast Support Multiple Subscriptions</a>

- Declares `multiple_subscriptions` gateway support so Payfast is eligible for carts containing multiple subscriptions.
- Adds a second subscription product (`second-subscription-product`) to the e2e test initialization script.
- Adds a new e2e test (`multiple-subscriptions-payment.test.js`) covering both block checkout and shortcode checkout flows, asserting that after payment completes the order is in `processing` state and both subscriptions are `Active`.

### Steps to test the changes in this Pull Request:

1. Add this branch to a subscriptions site.
2. Attempt to purchase multiple subscriptions in one order via both the block checkout and the shortcode checkout.
3. Verify the order is in processing state and both subscriptions are active.
4. Run the e2e test suite and confirm `multiple-subscriptions-payment.test.js` passes.

### Changelog entry


> Add -  Added support for multiple subscriptions purchased in one order.

<!-- START COPILOT CODING AGENT TIPS -->
---

💬 Send tasks to Copilot coding agent from [Slack](https://gh.io/cca-slack-docs) and [Teams](https://gh.io/cca-teams-docs) to turn conversations into code. Copilot posts an update in your thread when it's finished.

Closes #377

ryanr14 committed:
Update class-wc-gateway-payfast.php

Adds the required `multiple_subscriptions` feature support, so merchants can purchase multiple subscriptions in one order via PayFast.

ryanr14 committed:
Update changelog.txt

Recorded a new potential version with support for multiple subscriptions.

ryanr14 committed:
Update woocommerce-gateway-payfast.php

Version bump

dkotter commented on changelog.txt:
```diff
@@ -1,5 +1,8 @@
 *** PayFast Changelog ***
 
+= 1.7.1 - 2025-mm-dd =
+* Add - Added support for multiple subscriptions purchased in one order.
```
Changelog is updated when we put together the release PR so no need to update that here

dkotter commented on woocommerce-gateway-payfast.php:
```diff
@@ -6,7 +6,7 @@
  * Author: WooCommerce
  * Author URI: https://woocommerce.com/
- * Version: 1.7.0
+ * Version: 1.7.1
```
Version number will be bumped during the release process so this doesn't need updated
```suggestion
 * Version: 1.7.0
```

dkotter commented on includes/class-wc-gateway-payfast.php:
```diff
@@ -145,6 +145,7 @@ public function __construct() {
 			'subscription_date_changes',
 			'subscription_payment_method_change', // Subs 1.x support.
 			'subscription_payment_method_change_customer', // Enabled for https://github.com/woocommerce/woocommerce-gateway-payfast/issues/32.
+			'multiple_subscriptions',
```
I've not tested things but are there no specific updates needed to support multiple subscriptions, things just work as long as we declare support?

ryanr14 replied:
That is my understanding based on this doc section here, "[Step 1.1: Registering Support for Subscription Management Features](https://woocommerce.com/document/subscriptions/develop/payment-gateway-integration/#register-support)"

dkotter replied:
Reading that section, it shows how to _declare_ support in your extension for each of those subscription features. But we need to be sure multiple subscriptions actually work before we can declare that compatibility. Has testing been done to verify multiple subscriptions works here?

ryanr14 committed:
Delete changelog.txt

I had manually updated the changelog, not needed as the release process takes care of this.

ryanr14 committed:
Delete woocommerce-gateway-payfast.php

The version number increase is handled during the release process, I don't need to manually change that.

ryanr14 committed:
Updating files in my pull request to remove unneeded changes. No need to add changelog entries or change the plugin version number. This is all handled by the release process.

ryanr14 committed:
Update changelog.txt

ryanr14 committed:
Update woocommerce-gateway-payfast.php

jeffpaul commented:
@ryanr14 thanks for working on this here, please reach out once your updates and testing shows that multiple subscriptions works as expected and ideally that [existing Critical Flows](https://github.com/woocommerce/woocommerce-gateway-payfast/wiki/Critical-Flows) pass as well.

ryanr14 committed:
Merge branch 'trunk' into support-multiple-subscriptions

ryanr14 commented on includes/class-wc-gateway-payfast.php:
```diff
@@ -145,6 +145,7 @@ public function __construct() {
 			'subscription_date_changes',
 			'subscription_payment_method_change', // Subs 1.x support.
 			'subscription_payment_method_change_customer', // Enabled for https://github.com/woocommerce/woocommerce-gateway-payfast/issues/32.
+			'multiple_subscriptions',
 		);
```
@copilot apply changes based on [this feedback](https://github.com/woocommerce/woocommerce-gateway-payfast/pull/295#discussion_r3006558860)

Copilot committed:
Add e2e test for multiple subscriptions purchase

Agent-Logs-Url: https://github.com/woocommerce/woocommerce-gateway-payfast/sessions/07edd42a-6d0b-42c2-9b98-0527fd12764f

Co-authored-by: ryanr14 <867748+ryanr14@users.noreply.github.com>

iamdharmesh reviewed (commented):
Thanks for the PR @ryanr14. PR looks good overall. However, I wanted to resolve a few things before we proceed further on this.
- Pending Response [here](https://github.com/woocommerce/woocommerce-gateway-payfast/pull/295/changes#r2079920696)
- Regarding E2E tests, could you please add tests for the renewal as well?
- We may have to update [this](https://github.com/woocommerce/woocommerce-gateway-payfast/blob/trunk/includes/class-wc-gateway-payfast-privacy.php#L229) code to handle multiple subscriptions. 

Thanks.

ryanr14 committed:
Add grouped subscription renewal coverage

ryanr14 committed:
Guard shared Payfast subscription tokens

ryanr14 committed:
Fix subscription utility array alignment

iamdharmesh committed:
Merge branch 'trunk' into support-multiple-subscriptions

ryanr14 committed:
Stabilize multiple subscription renewal e2e

iamdharmesh commented on tests/e2e/specs/payment-flow/multiple-subscriptions-payment.test.js:
```diff
@@ -0,0 +1,219 @@
+		expect( subscriptionIdsForRenewal ).toHaveLength( 2 );
+
+		for ( const subscriptionId of subscriptionIdsForRenewal ) {
+			const orderId = await renewSubscriptionByCustomer( {page: checkoutPage, subscriptionId} );
```
I think better to renew this on the admin to make sure that token works correctly. renew subscription on customer side using payfast payment method will generate new payfast token.

ryanr14 replied:
Good call, thanks for catching that. I pushed a follow-up that moves this renewal coverage to the admin/scheduled renewal path instead of the customer `Renew now` flow.

The test now fakes the PayFast ad-hoc API request and asserts the request uses the existing `_payfast_subscription_token`, then confirms the token stays the same after the renewal ITN completes. I also kicked off a fresh E2E run on the updated head and it passed: https://github.com/woocommerce/woocommerce-gateway-payfast/actions/runs/26336175651

iamdharmesh reviewed (commented):
Thanks for the changes @ryanr14. PR looks good now. I only have one note on E2E test case. Also, Could you please help with checking on failing E2E tests?

ryanr14 committed:
Test admin renewal reuses Payfast token

ryanr14 committed:
Fix e2e helper coding standards

iamdharmesh reviewed (approved):
Thanks @ryanr14. Looks good to me.

iamdharmesh committed:
Merge branch 'trunk' into support-multiple-subscriptions

ryanr14 committed:
Fix plugin check workflow setup

ryanr14 committed:
Add direct access guard to Payfast privacy include

iamdharmesh committed:
Merge branch 'trunk' into support-multiple-subscriptions

iamdharmesh committed:
E2E: Fix typo: corrected 'passPharse' to 'passPhrase'.

iamdharmesh committed:
E2E: Update subscriptions option.

iamdharmesh committed:
Reverted plugin check workflow file changes.
